interfaces: T7268: Add op-mode command for show all interfaces on system

l0crian1 opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change summary
<!--- Provide a general summary of your changes in the Title above -->
This change adds a command to show all interfaces on the host system; not just those that are configurable by VyOS. This will allow the API to pull interfaces for dashboards that have been created by additionally installed packages or containers.

Syntax:
`show interfaces kernel`
`show interfaces kernel detail`
`show interfaces kernel json`

`show interfaces kernel <interface>`
`show interfaces kernel <interface> detail`
`show interfaces kernel <interface> json`

### `show interfaces kernel` (notice non-VyOS interfaces present)
```
Interface    IP Address          MAC                VRF        MTU  S/L    Description
-----------  ------------------  -----------------  -------  -----  -----  -------------
dum0         10.0.0.4/32         52:58:74:b1:69:69  default   1500  u/u
dum55        10.55.1.1/32        3e:62:ab:db:b9:31  test      1500  u/u
eth0         172.16.0.6/28       f4:52:14:cf:b7:27  default   1500  u/u    LAN Segments
eth1         -                   00:f0:cb:ef:de:55  default   1500  u/u
eth2         -                   00:f0:cb:ef:de:57  default   1500  u/u
eth3         10.55.0.1/24        00:f0:cb:ef:de:56  default   1500  A/D
             fd00::1/64
eth4         -                   f4:52:14:cf:b7:26  default   1500  u/D
eth5         10.243.100.1/16     d2:40:b0:6f:ef:ab  default   2800  u/u
eth6         10.244.249.28/16    6e:74:7d:bb:b7:6f  default   2800  u/u
lo           127.0.0.1/8         00:00:00:00:00:00  default  65536  u/u
             ::1/128
pim6reg      -                                      default   1452  u/u
tailscale0   100.113.245.121/32                     default   1280  u/u
wlan0        -                   06:33:c2:d1:14:40  default   1500  u/D
ztmosonlkl   10.244.100.10/24    02:45:3d:33:42:89  default   2800  u/u
```
###  `show interfaces kernel eth0 detail`
```
 Interface               | eth0
 IP Address              | 172.16.0.6/28
 MAC                     | f4:52:14:cf:b7:27
 VRF                     | default
 MTU                     | 1500
 S/L                     | u/u
 Description             | LAN Segments
 Device                  | Mellanox Technologies MT27500 Family [ConnectX-3] [15b3:1003]
 Alternate Names         | enp4s0d1
 Minimum MTU             | 68
 Maximum MTU             | 9900
 RX_Packets              | 71097024
 RX_Bytes                | 82417645611
 RX_Errors               | 0
 RX_Dropped              | 244758
 Receive Overrun Errors  | 0
 Received Multicast      | 9618172
 TX_Packets              | 119031343
 TX_Bytes                | 88422185319
 TX_Errors               | 0
 TX_Dropped              | 0
 Transmit Carrier Errors | 0
 Transmit Collisions     | 0
```
## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- optional: Link to related other tasks on Phabricator. -->
<!-- * https://vyos.dev/Txxxx -->
https://vyos.dev/T7268
## Related PR(s)
<!-- Link here any PRs in other repositories that are required by this PR -->

## How to test / Smoketest result
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```

Or provide the output of the smoketest

```
$ /usr/libexec/vyos/tests/smoke/cli/test_xxx_feature.py
test_01_simple_options (__main__.TestFeature.test_01_simple_options) ... ok
```
-->
`show interfaces kernel`
`show interfaces kernel detail`
`show interfaces kernel json`
`show interfaces kernel <interface>`
`show interfaces kernel <interface> detail`
`show interfaces kernel <interface> json`

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

l0crian1 committed:
interfaces: T7268: Add op-mode command for show all interfaces on system
This creates a command to show all interfaces on the system in both human readable and JSON format.

This includes non-VyOS interfaces like those created by containers.

Syntax:
show interfaces kernel
show interfaces kernel json

show interfaces kernel <interface>
show interfaces kernel <interface> json

sarthurdev reviewed (commented):
@dmbaturin can confirm, but I believe all new op-mode should be using the new standard in a `src/op_mode` script so that it can also be called via API.

l0crian1 commented:
> @dmbaturin can confirm, but I believe all new op-mode should be using the new standard in a `src/op_mode` script so that it can also be called via API.

This can be called by REST, just not GraphQL. I was actually targeting this command for the REST calls specifically.

c-po commented on op-mode-definitions/show-interfaces.xml.in:
```diff
@@ -26,6 +26,36 @@
+                <script>ip -j link show | jq -r '.[].ifname'</script>
+              </completionHelp>
+            </properties>
+            <command>sudo ip -s link show dev $4</command>
```
I see no difference in calling this with and without sudo?

l0crian1 replied:
Thanks! I have a bad habit of defaulting to sudo. I'll make sure it is correct in the final commit.

c-po reviewed (changes requested):
I tend to agree with @sarthurdev. It might look overcomplicated wrapping those calls in a python script. But the new design is clear, new op-mode commands should go in using the op-mode command wrapper.

l0crian1 commented:
@c-po @sarthurdev 
> I tend to agree with @sarthurdev. It might look overcomplicated wrapping those calls in a python script. But the new design is clear, new op-mode commands should go in using the op-mode command wrapper.

No worries, I'll go ahead and make a function for it in `interfaces.py`. Since it'll now go in a python script, do you want it to be in a table format like the normal`show interfaces`? Or keep it as its native output?

Another question on the same topic. Is it fine having just the command in the `completionHelp` section? Or do you want that in an actual script as well?

l0crian1 committed:
t7268: Add op-mode command for show all interfaces on system

Created op-mode script per request

Commands added:

show interfaces kernel
show interfaces kernel detail
show interfaces kernel json

show interfaces kernel <interface>
show interfaces kernel <interface> detail
show interfaces kernel <interface> json

l0crian1 committed:
interfaces: t7268: Add op-mode command for show all interfaces on system

Fixed XML formatting.

l0crian1 committed:
t7268: Add op-mode command for show all interfaces on system

lo address was an edge case and needed to be handled.

l0crian1 commented:
@sarthurdev @c-po 

Integrated functionality into `interfaces.py`. Commands available are:
```
show interfaces kernel
show interfaces kernel detail
show interfaces kernel json

show interfaces kernel <interface>
show interfaces kernel <interface> detail
show interfaces kernel <interface> json
```

Output looks like this:
### `show interfaces kernel` (notice non-VyOS interfaces present)
```
Interface    IP Address          MAC                VRF        MTU  S/L    Description
-----------  ------------------  -----------------  -------  -----  -----  -------------
dum0         10.0.0.4/32         52:58:74:b1:69:69  default   1500  u/u
dum55        10.55.1.1/32        3e:62:ab:db:b9:31  test      1500  u/u
eth0         172.16.0.6/28       f4:52:14:cf:b7:27  default   1500  u/u    LAN Segments
eth1         -                   00:f0:cb:ef:de:55  default   1500  u/u
eth2         -                   00:f0:cb:ef:de:57  default   1500  u/u
eth3         10.55.0.1/24        00:f0:cb:ef:de:56  default   1500  A/D
             fd00::1/64
eth4         -                   f4:52:14:cf:b7:26  default   1500  u/D
eth5         10.243.100.1/16     d2:40:b0:6f:ef:ab  default   2800  u/u
eth6         10.244.249.28/16    6e:74:7d:bb:b7:6f  default   2800  u/u
lo           127.0.0.1/8         00:00:00:00:00:00  default  65536  u/u
             ::1/128
pim6reg      -                                      default   1452  u/u
tailscale0   100.113.245.121/32                     default   1280  u/u
wlan0        -                   06:33:c2:d1:14:40  default   1500  u/D
ztmosonlkl   10.244.100.10/24    02:45:3d:33:42:89  default   2800  u/u
```
###  `show interfaces kernel eth0 detail`
```
 Interface               | eth0
 IP Address              | 172.16.0.6/28
 MAC                     | f4:52:14:cf:b7:27
 VRF                     | default
 MTU                     | 1500
 S/L                     | u/u
 Description             | LAN Segments
 Device                  | Mellanox Technologies MT27500 Family [ConnectX-3] [15b3:1003]
 Alternate Names         | enp4s0d1
 Minimum MTU             | 68
 Maximum MTU             | 9900
 RX_Packets              | 71097024
 RX_Bytes                | 82417645611
 RX_Errors               | 0
 RX_Dropped              | 244758
 Receive Overrun Errors  | 0
 Received Multicast      | 9618172
 TX_Packets              | 119031343
 TX_Bytes                | 88422185319
 TX_Errors               | 0
 TX_Dropped              | 0
 Transmit Carrier Errors | 0
 Transmit Collisions     | 0
```

dmbaturin commented on src/op_mode/interfaces.py:
```diff
@@ -296,6 +305,114 @@ def _get_counter_data(ifname: typing.Optional[str],
+    else:
+        int_name = ''
+
+    kernel_interface = json.loads(cmd(f'ip -j -d -s address show{int_name}'))
```
```suggestion
    kernel_interface = json.loads(cmd(f'ip -j -d -s address show {int_name}'))
```

A space embedded into a value and omitted in the format string `show{int_name}` is an easily avoidable footgun, in my opinion. Someone edits the `int_name` assignment and the command will become invalid. A space in the format string makes any edits in `int_name` safe.

l0crian1 replied:
Thanks, I've always gone back and forth on what is correct there. Good to have clarification.

dmbaturin commented on src/op_mode/interfaces.py:
```diff
@@ -296,6 +305,114 @@ def _get_counter_data(ifname: typing.Optional[str],
+        # Check if the interface exists
+        if not interface_exists(ifname):
+            raise vyos.opmode.Error(f"{ifname} does not exist!")
+        int_name = f' dev {ifname}'
```
```suggestion
        int_name = f'dev {ifname}'
```

dmbaturin commented on src/op_mode/interfaces.py:
```diff
@@ -296,6 +305,114 @@ def _get_counter_data(ifname: typing.Optional[str],
+    if ifname:
+        # Check if the interface exists
+        if not interface_exists(ifname):
+            raise vyos.opmode.Error(f"{ifname} does not exist!")
```
```suggestion
            raise vyos.opmode.IncorrectValue(f"{ifname} does not exist!")
```

`vyos.opmode.Error` is the base class, actual errors should be as specific as possible. There are two errors that may apply here: `UnconfiguredObject` and `IncorrectValue`. I'm leaning towards `IncorrectValue` because in this case interfaces may not exist in the _configuration_ but still exist in the system.

l0crian1 replied:
Thanks, change made!

l0crian1 committed:
interfaces: T7268: Add op-mode command for show all interfaces on system

Commiting suggestions from dmbaturin

Co-authored-by: Daniil Baturin <daniil@baturin.org>

jestabro reviewed (approved):
All works as expected, including REST and GraphQL requests.

dmbaturin reviewed (approved):
Change requests are addressed, it's now a proper part of the op mode structure. No more objections, thanks!
